packages/flutter/test/widgets/router_restoration_test.dart

```diff
@@ -0,0 +1,167 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  testWidgets('Router state restoration without RouteInfomrationProvider', (WidgetTester tester) async {
+    final UniqueKey router = UniqueKey();
+    _TestRouterDelegate delegate() => tester.widget<Router<Object?>>(find.byKey(router)).routerDelegate as _TestRouterDelegate;
+
+    await tester.pumpWidget(_TestWidget(routerKey: router));
+    expect(find.text('Current config: null'), findsOneWidget);
+    expect(delegate().newRoutePaths, isEmpty);
+    expect(delegate().restoredRoutePaths, isEmpty);
+
+    delegate().currentConfiguration = '/foo';
+    await tester.pumpAndSettle();
+    expect(find.text('Current config: /foo'), findsOneWidget);
+    expect(delegate().newRoutePaths, isEmpty);
+    expect(delegate().restoredRoutePaths, isEmpty);
+
+    await tester.restartAndRestore();
```

Comment:
I didn't realize someone built this state restoration stuff into tester, this is great that we can test this 👍